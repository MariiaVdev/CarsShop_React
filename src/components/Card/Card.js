import React, { useContext } from 'react';
import styles from './Card.module.scss';
import Button from '../Button/Button';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { setIsOpenModalAC, setModalDataAC } from '../../store/modal/actionCreators';
import { setCardDataAC } from '../../store/card/actionCreators';
import { markAsFavoriteAC, countFavoriteAC } from '../../store/cars/actionCreators';
import ViewContext from '../../contexts/viewContext/ViewContext';


const Card = (props) => {
    const {  name, price, img, code, color, isFavorite } = props;
    const dispatch = useDispatch();
    const {isCardThem} = useContext(ViewContext);

    return (
        <>
        {isCardThem ? 
        <div className={styles.cardWrapper}>
            <img src={img} alt={name}></img>
            <div className={styles.cardName}><h3>{name}
                <button className={styles.cardFavorite} onClick={() => {
                    dispatch(markAsFavoriteAC(code));
                    dispatch(countFavoriteAC());
                }}>
                    {isFavorite
                        ? <span className={styles.cardFavoriteSelect}></span>
                        : <span className={styles.cardFavoriteNoSelect}></span>
                    }
                </button>
            </h3></div>
            <div className={styles.cardCode}>Vendor code: {code}</div>
            <div className={styles.cardCode}>Color: <button className={styles.cardBtnColor} style={{ backgroundColor: color }}></button></div>
            <div className={styles.cardPrice}><p><b>${price}</b>
                <Button bgColor={'black'} textBtn={'Add to cart'} onClick={() => {
                    dispatch(setModalDataAC({ code, name }));
                    dispatch(setCardDataAC({ name, price, img, code, color }))
                    dispatch(setIsOpenModalAC(true));
                }} /></p>
            </div>
        </div>
        
        : 
            <tr>
                <td>
        <img className={styles.tblImg} src={img} alt={name}></img>
        </td>
        
        <td>
        <h3 className={styles.tblCardName}>{name} </h3>
        </td>
        <td>
            <button className={styles.cardFavorite} onClick={() => {
                dispatch(markAsFavoriteAC(code));
                dispatch(countFavoriteAC());
            }}>
                {isFavorite
                    ? <span className={styles.cardFavoriteSelect}></span>
                    : <span className={styles.cardFavoriteNoSelect}></span>
                }
            </button>
       
        </td>
        <td className={styles.tblCardCode}><b>Vendor code:</b> {code}</td>
        <td className={styles.tblCardCode}><b>Color:</b></td><td> <button className={styles.tblCardBtnColor} style={{ backgroundColor: color }}></button></td>
        <td className={styles.tblCardPrice}><p><b>${price}</b></p> </td>
        <td>
            <Button bgColor={'black'} textBtn={'Add to cart'} onClick={() => {
                dispatch(setModalDataAC({ code, name }));
                dispatch(setCardDataAC({ name, price, img, code, color }))
                dispatch(setIsOpenModalAC(true));
            }} />
            </td>  
        </tr>
    }
        </>
    )
}

Card.propTypes = {
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    img: PropTypes.string.isRequired,
    code: PropTypes.number,
    color: PropTypes.string,
    isFavorite: PropTypes.bool.isRequired
}

Card.defaultProps = {
    code: 1234,
    color: 'grey'
}

export default Card;