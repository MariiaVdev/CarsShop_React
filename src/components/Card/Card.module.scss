.cardWrapper {
  display: flex;
  width: 400px;
  flex-direction: column;
  box-sizing: border-box;
  padding: 14px;
  margin: 20px 20px;
  border: 1px solid lightgrey;
  border-radius: 5px;

  img {
    width: 367px;
    height: 200px;
  }

  .cardName {
    font-size: 22px;
  }

  .cardCode {
    font-size: 18px;
    padding-bottom: 5px;
  }

  .cardBtnColor {
    width: 20px;
    height: 20px;
    border: 1px lightgrey solid;

  }

  .cardPrice p {
    margin-top: 10px;
    margin-bottom: 10px;
    font-size: 28px;
  }
}

.cardFavorite {
  background-color: transparent;
  border: none;
  background-repeat: no-repeat;
  background-position: center center;
  float: right;

  .cardFavoriteNoSelect {
    width: 28px;
    height: 28px;
    background-repeat: no-repeat;
    background-position: center center;
    float: right;

    background-image: url("../../../public/img/favorite_Not.png");
  }

  .cardFavoriteSelect {
    width: 28px;
    height: 28px;
    background-repeat: no-repeat;
    background-position: center center;
    float: right;

    background-image: url("../../../public/img/favorite_Yes.png");
  }
}

.tblImg {
  width: 367px;
  height: 200px;
}

tr {
  width: 100%;
}

td {
  width: auto;
  text-align: center;
  padding-left: 10px;
  padding-right: 10px;
}

.tblCardCode {
  font-size: 22px;
  padding-bottom: 5px;
}

.tblCardName {
  font-size: 30px;
}

.tblCardPrice {
  margin-top: 10px;
  margin-bottom: 10px;
  font-size: 28px;
}

.tblCardBtnColor {
  width: 30px;
  height: 30px;
  border: 1px lightgrey solid;
}