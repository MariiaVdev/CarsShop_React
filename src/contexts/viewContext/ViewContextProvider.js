import { useState } from "react";
import ViewContext from "./ViewContext";

const ViewContextProvider = (props) => {
    const { children } = props;

    const [isCardThem, setIsCardTheme] = useState(true);

    return (
        <ViewContext.Provider value={{ isCardThem, setIsCardTheme }}>
            {children}
        </ViewContext.Provider>
    )
}

export default ViewContextProvider;