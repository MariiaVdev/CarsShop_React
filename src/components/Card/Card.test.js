import Card from './Card'
import renderWithProviders from '../../utils/utilsModalForm/utilsModalFormTest';
import ViewContextProvider from '../../contexts/viewContext/ViewContextProvider';
import UtilsViewContextProvider from '../../utils/utilsCard/utilsViewContextProvider'


const ComponentCard = () => {
    return (
        <ViewContextProvider>
            <Card name={'test'} price={12345} img={'img'} isFavorite={false} />
        </ViewContextProvider>
    )
}
const ComponentTable = () => {
    return (
        <UtilsViewContextProvider>
            <Card name={'test'} price={12345} img={'img'} code={1234} color={'red'} isFavorite={true} />
        </UtilsViewContextProvider>
    )
}


describe("Render Card", () => {
    test("should Card render in Card view, default props and Favorite is false", () => {
        const { asFragment } = renderWithProviders(<ComponentCard />);
        expect(asFragment()).toMatchSnapshot()
    })
    test("should Card render in Table view and Favorite is true", () => {
        const { asFragment } = renderWithProviders(<ComponentTable />);
        expect(asFragment()).toMatchSnapshot()
    })
})


