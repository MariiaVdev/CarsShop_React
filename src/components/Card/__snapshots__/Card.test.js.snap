// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Card should Card render in Card view, default props and Favorite is false 1`] = `
<DocumentFragment>
  <div
    class="cardWrapper"
  >
    <img
      alt="test"
      src="img"
    />
    <div
      class="cardName"
    >
      <h3>
        test
        <button
          class="cardFavorite"
        >
          <span
            class="cardFavoriteNoSelect"
          />
        </button>
      </h3>
    </div>
    <div
      class="cardCode"
    >
      Vendor code: 1234
    </div>
    <div
      class="cardCode"
    >
      Color: 
      <button
        class="cardBtnColor"
        style="background-color: grey;"
      />
    </div>
    <div
      class="cardPrice"
    >
      <p>
        <b>
          $12345
        </b>
        <button
          class="btn"
          style="background-color: black;"
        >
          Add to cart
        </button>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Render Card should Card render in Table view and Favorite is true 1`] = `
<DocumentFragment>
  <img
    alt="test"
    class="tblImg"
    src="img"
  />
  <h3
    class="tblCardName"
  >
    test 
  </h3>
  <button
    class="cardFavorite"
  >
    <span
      class="cardFavoriteSelect"
    />
  </button>
  <b>
    Vendor code:
  </b>
   1234
  <b>
    Color:
  </b>
   
  <button
    class="tblCardBtnColor"
    style="background-color: red;"
  />
  <p>
    <b>
      $12345
    </b>
  </p>
   
  <button
    class="btn"
    style="background-color: black;"
  >
    Add to cart
  </button>
</DocumentFragment>
`;
